apps/blockchain/lib/blockchain/ethash.ex

```diff
@@ -0,0 +1,56 @@
+defmodule Blockchain.Ethash do
+  @moduledoc """
+  This module contains the logic found in Appendix J of the
+  yellow paper concerning the Ethash implementation for POW.
+  """
+
+  @j_epoch 30_000
+  @j_datasetinit :math.pow(2, 30) |> round
+  @j_datasetgrowth :math.pow(2, 23) |> round
+  @j_mixbytes 128
+  @j_cacheinit :math.pow(2, 24) |> round
+  @j_cachegrowth :math.pow(2, 17) |> round
+  @j_hashbytes 64
+
+  def epoch(block_number) do
+    div(block_number, @j_epoch)
+  end
+
+  def dataset_size(epoch, cache \\ Ethash.Cache.data_sizes()) do
+    case Enum.at(cache, epoch) do
+      nil -> calculate_dataset_size(epoch)
+      found_data_size -> found_data_size
+    end
+  end
+
+  defp calculate_dataset_size(epoch) do
+    prime(@j_datasetinit + @j_datasetgrowth * epoch - @j_mixbytes, @j_mixbytes)
+  end
+
+  def cache_size(epoch, cache \\ Ethash.Cache.cache_sizes()) do
+    case Enum.at(cache, epoch) do
+      nil -> calculate_cache_size(epoch)
+      found_cache -> found_cache
+    end
+  end
+
+  defp calculate_cache_size(epoch) do
+    prime(@j_cacheinit + @j_cachegrowth * epoch - @j_hashbytes, @j_hashbytes)
+  end
+
+  defp prime(x, y) do
```

Comment:
Is there a name that would be more descriptive then `prime`? Is this `greatest_common_factor`? Or `greatest_common_prime` or something?

Reply:
from the document, they state:

> we always take the highest prime below the linearly growing threshold

Reply:
Yeah, great idea. I'll update this to something more descriptive.

Reply:
@masonforest @hayesgm we just renamed it, but I'm not actually sure it's that much clearer. The difficult aspect of this is that the function doesn't actually return the highest prime number. It returns the largest size below the upper bound for which `x/y` is a prime number. Do you have any suggestions for other names or any thoughts on this name?

Reply:
I think I like `prime` over `largest_size`.

Other ideas:

`max_prime_quotient`?
`max_prime`?
`upper_bound_prime`?

Reply:
Unfortunately, `prime` is not actually a descriptive name because that's not what it returns. It actually returns numbers that aren't prime. It's just that the `x/y` has to be prime. it's like `x` needs to be prime _relative_ to a `y` factor or something like that.

Reply:
I mean, `largest_bounded_relative_prime` with a comment above it describing what you just wrote would be fine. It's definitely not an intuitive function at all (as with many Ethereum functions...)